### ISSUE TITLE
Place model formula into its archive

### ISSUE THREAD

**Issue by inferrna** [tensorflow/models#830]
my problem is very similar to https://github.com/tensorflow/models/issues/538 - I can't produce model with morphology-map and char-ngram-map files present. But the fact that pretrained models (https://github.com/tensorflow/models/blob/master/syntaxnet/universal.md) contains this files points to existence of the right formula. Can you, please, place it (train.sh, context.pbtx, etc) into each model's zip archive?

**Comment by xtknight** [comment]
I also have the same problem. There's no documentation or indication of where morphology-map, char-ngram-map, morph-label-set actually come from. Additionally there is no documentation about how to train the tokenizer or morpher modules.

**Comment by inferrna** [comment]
@xtknight thank you, that was helpful. I also created the gist from your code with patch for lexicon_builder.cc https://gist.github.com/inferrna/c8d22af97e03db5855372b55397024a2

**Comment by xtknight** [comment]
@inferrna Oh I see! Thanks.

Also, for char-ngram this might work. Well, all I did was copy logic from other pieces of the code. Seems to do the job.

    #include "syntaxnet/segmenter_utils.h"

...

    TermFrequencyMap char_ngram_map;
    int max_char_ngram_length = task_context_.Get("lexicon_max_char_ngram_length", 3);
    bool use_terminators = task_context_.Get("lexicon_char_ngram_include_terminators", false);

... (below part goes in same loop as the ones in my previous comment) ...

        vector<tensorflow::StringPiece> char_spN;
        if (use_terminators) char_spN.push_back("^");
        SegmenterUtils::GetUTF8Chars(token.word(), &char_spN);
        if (use_terminators) char_spN.push_back("$");
        for (int start = 0; start < char_spN.size(); ++start) {
          string char_ngram;
          for (int index = 0;
               index < max_char_ngram_length && start + index < char_spN.size();
               ++index) {
            tensorflow::StringPiece c = char_spN[start + index];
            if (c == " ") break;  // Never add char ngrams containing spaces.
            tensorflow::strings::StrAppend(&char_ngram, c);

            char_ngram_map.Increment(char_ngram);
          }
        }

...


    char_ngram_map.Save(TaskContext::InputFile(*task_context_.GetInput("char-ngram-map")));


Rather unrelated, but the confusing part is that lexicon builder produces only chars in the word-map when training the tokenizer. Whereas when training for parsing, tagging it produces words in word-map. Then there's a last-word-map in some of the Universal Dependencies files that appears to be same as word-map. Well, it would be great if someone could clarify once and for all what these files are and where they came from. I wonder if they're using some different internal version of SyntaxNet.

**Comment by inferrna** [comment]
@xtknight thanks, this also working for me. Also I updated the gist.
Is there any docs describing these files (char-ngram-map, morph-label-set, morphology-map) ?
It seems, that you know more than syntaxnet developers (they still keep silent here and in other related threads).

**Comment by xtknight** [comment]
@inferrna 

I have actually been working on a guide here. But it's a long way from being complete. I wasn't actually going to post it until it's more organized, but then again, I don't know when that would be.

https://docs.google.com/document/d/1dPMkyzh7-QeSk7ONUuMTj9lY6flQ0NXdWcYzcMtG2oc/edit?usp=sharing

The "Appendix" section is a mess, but before that there's some info on how things get embedded and I briefly go over the structure of some of those maps and tables.

Yes, they don't seem to be big on documentation but there are some useful comments and especially test suites that really helped me figure out what's going on. It wasn't actually as hard as I thought it would be. It's one of those situations where you need to read the code first and then read their tutorial, and then their tutorial on the git page makes a lot more sense.

The most confusing part is that tokenization, morphing, tagging, and parsing are almost all treated the exact same way, except for the transition system itself (which amounts to the output of the neural network). Anyway, I go over it in the guide.

I've also just gotten around to adding my own word lexical category word features (like punctuation-amount but it's a specific feature for the Korean language that dissects the lead, vowel, and tail from the character). I'll add it to the guide sooner or later.

As much as I've dissected the original components, I'll be honest...I still have no idea how the entire thing is really supposed to work, especially for morphologically complex languages like Japanese and Korean. They seem to omit accuracy of Japanese from the table even though it's in Universal Dependencies and I wonder if there's a reason for that.

**Comment by inferrna** [comment]
@xtknight May be you also know how to get morpher-params file? I'm also looked on https://github.com/dsindex/syntaxnet/ - they are using it in parse.sh but no words about generating it in train.sh, just parser-params and tagger-params.

**Comment by xtknight** [comment]
morpher-params should be the trained model file of morpher renamed to that name. I only have basic information about training the morpher in my documentation in the appendix but to be honest I have very low confidence about it. I ended up not using the morpher but

When I did use it here's what I did (roughly):

    MORPH_HIDDEN_LAYER_SIZES=128
    MORPH_HIDDEN_LAYER_PARAMS=128
    GREEDY_MORPH_PARAMS=${MORPH_HIDDEN_LAYER_PARAMS}-0.08-3600-0.9-0

    function train_morpher {
	    ${BINDIR}/parser_trainer \
	      --task_context=${TMP_DIR}/context \
	      --arg_prefix=brain_morpher \
	      --compute_lexicon \
	      --graph_builder=greedy \
	      --training_corpus=tagged-training-corpus \
	      --tuning_corpus=tagged-tuning-corpus \
	      --output_path=${TMP_DIR} \
	      --batch_size=${BATCH_SIZE} \
	      --decay_steps=3600 \
	      --hidden_layer_sizes=${MORPH_HIDDEN_LAYER_SIZES} \
	      --learning_rate=0.08 \
	      --momentum=0.9 \
	      --beam_size=1 \
	      --seed=0 \
	      --params=${GREEDY_MORPH_PARAMS} \
	      --num_epochs=12 \
	      --report_every=10 \
	      --checkpoint_every=100 \
	      --logtostderr
    }


Training data became something like this (very roughly, it should be tenses and aspects instead of what I did with TagFirst/TagLast, but anyways it just trains against this list of parameters per token):

    1	프랑스의	_	NOUN	NOUN	TagFirst=NNP|TagLast=JKG	_	NP_MOD	4
    2	세계적인	_	COP	COP	TagFirst=NNG|Tag1=XSN|Tag2=VCP|TagLast=ETM	_	VNP_MOD	4

I'm not sure if this can be useful for you.

**Comment by WenchenLi** [comment]
thank you guys @inferrna @xtknight for making training UD datasets working, however it seems adding n-gram and morph features not helps with the parser result on UD Chinese, (always around 68%), has anybody experienced the same situation here?

**Comment by inferrna** [comment]
Just found some good news from google:
https://research.googleblog.com/2017/03/an-upgrade-to-syntaxnet-new-models-and.html
https://github.com/tensorflow/models/blob/master/syntaxnet/g3doc/conll2017/README.md
http://universaldependencies.org/conll17/

It seems like the truth is out there, but not all in one place

**Comment by karmel** [comment]
@calberti Can you comment?

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
